engines/wintermute/base/gfx/opengl/meshx.cpp

```diff
@@ -0,0 +1,498 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+/*
+ * This file is based on WME.
+ * http://dead-code.org/redir.php?target=wme
+ * Copyright (c) 2003-2013 Jan Nedoma and contributors
+ */
+
+#include "common/math.h"
+#include "engines/wintermute/base/base_game.h"
+#include "engines/wintermute/base/gfx/opengl/base_surface_opengl3d.h"
+#include "engines/wintermute/base/gfx/opengl/material.h"
+#include "engines/wintermute/base/gfx/opengl/meshx.h"
+#include "engines/wintermute/base/gfx/opengl/shadow_volume.h"
+#include "engines/wintermute/base/gfx/x/frame_node.h"
+#include "engines/wintermute/base/gfx/x/loader_x.h"
+#include "engines/wintermute/base/gfx/x/modelx.h"
+#include "engines/wintermute/dcgf.h"
+
+namespace Wintermute {
+
+//////////////////////////////////////////////////////////////////////////
+MeshX::MeshX(BaseGame *inGame) : BaseNamedObject(inGame),
+                                 _BBoxStart(0.0f, 0.0f, 0.0f), _BBoxEnd(0.0f, 0.0f, 0.0f),
+                                 _numAttrs(0), _maxFaceInfluence(0),
+                                 _skinAdjacency(nullptr), _adjacency(nullptr), _skinnedMesh(false) {
+}
+
+//////////////////////////////////////////////////////////////////////////
+MeshX::~MeshX() {
+	delete[] _adjacency;
+	delete[] _skinAdjacency;
+	delete[] _vertexPositionData;
+
+	for (uint32 i = 0; i < _materials.size(); i++) {
+		delete _materials[i];
+	}
+
+	_materials.clear();
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::loadFromX(const Common::String &filename, XFileLexer &lexer) {
+	bool res = true;
+
+	lexer.advanceToNextToken(); // skip the name
+	lexer.advanceOnOpenBraces();
+	_vertexCount = readInt(lexer);
+
+	// vertex format for .X meshes will be position + normals + textures
+	const int vertexSize = 32;
+	_vertexData = new byte[vertexSize * _vertexCount];
+	_vertexPositionData = new float[3 * _vertexCount];
+	float *vertexCoords = reinterpret_cast<float *>(_vertexData);
+
+	parsePositionCoords(lexer, vertexCoords);
+
+	int faceCount = readInt(lexer);
+	// we should be able to assume for now that
+	// we are only dealing with triangles
+	_indexData = new uint16[faceCount * 3];
+	_indexCount = faceCount * 3;
+
+	parseFaces(lexer, faceCount);
+
+	while (!lexer.eof()) {
+		if (lexer.tokenIsIdentifier("MeshTextureCoords")) {
+			lexer.advanceToNextToken();
+			lexer.advanceOnOpenBraces();
+
+			parseTextureCoords(lexer, vertexCoords);
+		} else if (lexer.tokenIsIdentifier("MeshNormals")) {
+			lexer.advanceToNextToken();
+			lexer.advanceOnOpenBraces();
+
+			parseNormalCoords(lexer, vertexCoords);
+		} else if (lexer.tokenIsIdentifier("MeshMaterialList")) {
+			lexer.advanceToNextToken();
+			lexer.advanceOnOpenBraces();
+
+			parseMaterials(lexer, faceCount, filename);
+		} else if (lexer.tokenIsIdentifier("Material")) {
+			warning("MeshX::loadFromX global material will not be properly handled");

```

Comment:
Why not?

Reply:
When I first wrote this, I was concerned about a non empty material list and so postponed this. However, if there is a global material, then there should be no material list, so we can use the global one for all faces. Will add implement proper support here.